src/dataset/impl_dataset.rs

```diff
@@ -724,50 +740,119 @@ where
-
         let mut objs: Vec<O> = Vec::new();
 
-        for i in 0..k {
-            assist_swap_array2(&mut records_sl, i, fold_size, features);
-            assist_swap_array2(&mut targets_sl, i, fold_size, targets);
-
-            let train = DatasetBase::new(
-                ArrayView2::from_shape(
-                    (samples_count - fold_size, features),
-                    records_sl.split_at(fold_size * features).1,
-                )
-                .unwrap(),
-                ArrayView2::from_shape(
-                    (samples_count - fold_size, targets),
-                    targets_sl.split_at(fold_size * targets).1,
-                )
-                .unwrap(),
-            );
-
-            let obj = fit_closure(train);
-            objs.push(obj);
-
-            assist_swap_array2(&mut records_sl, i, fold_size, features);
-            assist_swap_array2(&mut targets_sl, i, fold_size, targets);
+        {
+            let records_sl = self.records.as_slice_mut().unwrap();
+            let mut targets_sl2 = self.targets.as_multi_targets_mut();
+            let targets_sl = targets_sl2.as_slice_mut().unwrap();
+
+            for i in 0..k {
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+
+                {
+                    let train = DatasetBase::new(
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, features),
+                            records_sl.split_at(fold_size * features).1,
+                        )
+                        .unwrap(),
+                        ArrayView2::from_shape(
+                            (samples_count - fold_size, targets),
+                            targets_sl.split_at(fold_size * targets).1,
+                        )
+                        .unwrap(),
+                    );
+
+                    let obj = fit_closure(&train);
+                    objs.push(obj);
+                }
+
+                assist_swap_array2!(records_sl, i, fold_size, features);
+                assist_swap_array2!(targets_sl, i, fold_size, targets);
+            }
         }
 
         objs.into_iter().zip(self.sample_chunks(fold_size))
-        //
     }
-}
 
-fn assist_swap_array2<F>(slice: &mut [F], index: usize, fold_size: usize, features: usize) {
-    if index == 0 {
-        return;
+    pub fn cross_validate_mt<O, DT, ER: std::error::Error, M, C>(
+        &'a mut self,
+        k: usize,
+        parameters: &[M],
+        eval: C,
+    ) -> Array1<F>
+    where
+        DT: Data<Elem = E>,
+        M: for<'c> Fit<ArrayView2<'c, F>, ArrayView2<'c, E>, ER, Object = O>,
+        O: for<'d> PredictRef<ArrayView2<'a, F>, ArrayBase<DT, Ix2>>,
+        C: Fn(&ArrayBase<DT, Ix2>, &ArrayView2<E>) -> F,
+    {
+        let mut accuracies = Array1::from_elem(parameters.len(), F::zero());
```

Comment:
not necessarily `accuracies`, could be called `metrics` or `eval_results`

Reply:
Yes but this cross-validation proposal is still very early stage. I have in mind to return a struct with the set of fit results in order to be able to see, for example, the number of fits that returned an error and maybe on which parameters, and then add a method (maybe called `eval_with`) which takes a closure and only runs on successful fits. This is still very "cloudy" so I am not paying too much attention to the naming right now, but I agree on `eval_results` since one could compare them by loss instead of accuracy

Reply:
sounds like a good idea :+1: